components/speedreader/speedreader_loader.cc

```diff
@@ -0,0 +1,373 @@
+  // SpeedReaderURLLoader starts handling the request after
+  // OnReceivedResponse(). A redirect response is not expected.
+  NOTREACHED();
+}
+
+void SpeedReaderURLLoader::SetPriority(net::RequestPriority priority,
+                                       int32_t intra_priority_value) {
+  if (state_ == State::kAborted)
+    return;
+  source_url_loader_->SetPriority(priority, intra_priority_value);
+}
+
+void SpeedReaderURLLoader::PauseReadingBodyFromNet() {
+  if (state_ == State::kAborted)
+    return;
+  source_url_loader_->PauseReadingBodyFromNet();
+}
+
+void SpeedReaderURLLoader::ResumeReadingBodyFromNet() {
+  if (state_ == State::kAborted)
+    return;
+  source_url_loader_->ResumeReadingBodyFromNet();
+}
+
+void SpeedReaderURLLoader::OnBodyReadable(MojoResult) {
+  DCHECK_EQ(State::kLoading, state_);
+
+  size_t start_size = buffered_body_.size();
+  uint32_t read_bytes = kReadBufferSize;
+  buffered_body_.resize(start_size + read_bytes);
+  MojoResult result = body_consumer_handle_->ReadData(
+      &buffered_body_[0] + start_size, &read_bytes, MOJO_READ_DATA_FLAG_NONE);
+  switch (result) {
+    case MOJO_RESULT_OK:
+      break;
+    case MOJO_RESULT_FAILED_PRECONDITION:
+      // Reading is finished.
+      buffered_body_.resize(start_size);
+      MaybeLaunchSpeedreader();
+      return;
+    case MOJO_RESULT_SHOULD_WAIT:
+      body_consumer_watcher_.ArmOrNotify();
+      return;
+    default:
+      NOTREACHED();
+      return;
+  }
+
+  DCHECK_EQ(MOJO_RESULT_OK, result);
+  buffered_body_.resize(start_size + read_bytes);
+  // TODO(iefremov): We actually can partially |pumpContent| to speedreader,
+  // but skipping it for now to simplify things. Pumping is not free in terms
+  // of CPU ticks, so we will have to keep alive speedreader instance on another
+  // thread.
+
+  body_consumer_watcher_.ArmOrNotify();
+}
+
+void SpeedReaderURLLoader::OnBodyWritable(MojoResult r) {
+  DCHECK_EQ(State::kSending, state_);
+  if (bytes_remaining_in_buffer_ > 0) {
+    SendReceivedBodyToClient();
+  } else {
+    CompleteSending();
+  }
+}
+
+void SpeedReaderURLLoader::MaybeLaunchSpeedreader() {
+  DCHECK_EQ(State::kLoading, state_);
+  VLOG(2) << __func__ << " buffered body size = " << buffered_body_.size();
+
+  bytes_remaining_in_buffer_ = buffered_body_.size();
+  if (!throttle_) {
+    Abort();
+    return;
+  }
+
+  if (bytes_remaining_in_buffer_ > 0) {
+    // Offload heavy distilling to another thread.
+    base::PostTaskAndReplyWithResult(
+        FROM_HERE, {base::ThreadPool(), base::TaskPriority::USER_BLOCKING},
+        base::BindOnce(
+            [](GURL url, std::string data) -> auto {
+              SCOPED_UMA_HISTOGRAM_TIMER("Brave.Speedreader.Distill");
+              SpeedReader speedreader;
+              speedreader.reset(url.spec().c_str());
+              // TODO(iefremov): Change speedreader API to accept the data size?
+              data.push_back('\0');
+              speedreader.pumpContent(data.data());
```

Comment:
another TODO is to check that UTF  is handled correctly across the boundary. Rust strings by default handle UTF, but not 100% sure that moving that back and forth will be encoded correctly across all platforms

Reply:
@AndriusA can you create a ticket for this?

Reply:
so far I can tell it works fine for a bunch of Russian websites :D